### PR TITLE
Ensure if a docs render is torndown during preparation, it throws

### PR DESCRIPTION
**Pull request by tmeasday**
Issue: https://github.com/storybookjs/storybook/issues/19015

Implementing https://github.com/storybookjs/storybook/pull/17599 for DocsRenders

## What I did

Basically mirrored the behaviour of a story render in a docs render. The contract is, if you call `render.prepare()` and then `render.teardown()` before it's done, the prepare should reject with a `PREPARE_ABORTED`.

## How to test

- See integration tests in `XRender.test.ts`
- See integration tests in `PreviewWeb.test.ts`
- Run the `react-vite/default-js` sandbox and browser to http://localhost:6006. Previously it would error about 50% of the time on my machine.

### CODE REVIEW

File: code/lib/preview-web/src/render/StandaloneDocsRender.ts [hunk 1]
```diff
@@ -96,12 +98,17 @@ export class StandaloneDocsRender<TFramework extends AnyFramework> implements Re
     };
 
     this.rerender = async () => renderDocs();
-    this.teardown = async ({ viewModeChanged }: { viewModeChanged?: boolean } = {}) => {
+    this.teardownRender = async ({ viewModeChanged }: { viewModeChanged?: boolean } = {}) => {
       if (!viewModeChanged || !canvasElement) return;
```
Comment (shilman) [inline code/lib/preview-web/src/render/StandaloneDocsRender.ts:102]: does it need something like this?

```
if(this.torndown) return
```

Comment (shilman) [reply code/lib/preview-web/src/render/StandaloneDocsRender.ts:102]: or to throw an error?

**Review by shilman — COMMENTED** [review]
Mostly LGTM. I'm not sure what's going on with those jest changes tho

File: code/lib/preview-web/src/PreviewWeb.test.ts [hunk 2]
```diff
@@ -31,8 +31,8 @@ import { logger } from '@storybook/client-logger';
 import { addons, mockChannel as createMockChannel } from '@storybook/addons';
 import type { AnyFramework } from '@storybook/csf';
 import type { ModuleImportFn, WebProjectAnnotations } from '@storybook/store';
-import { expect } from '@jest/globals';
 import { mocked } from 'ts-jest/utils';
+import jestMock from 'jest-mock';
```
[the remark this answers is not in the document]
Comment (tmeasday) [reply code/lib/preview-web/src/PreviewWeb.test.ts:35]: Related to some jest changes: https://github.com/storybookjs/storybook/issues?q=is%3Apr+author%3Ashilman+sort%3Aupdated-desc. I'll revert this, not needed any more I guess.

File: code/lib/preview-web/src/render/StandaloneDocsRender.ts (hunk 1 above)
Comment (tmeasday) [reply code/lib/preview-web/src/render/StandaloneDocsRender.ts:102]: Hmm, we didn't have that before and we don't "need" it as we don't ever teardown >1. But throwing an error unexpected code paths isn't the worst thing to do. I'll add something.

Comment (tmeasday) [reply code/lib/preview-web/src/render/StandaloneDocsRender.ts:102]: Ok, I tried this and it turns out we do sometimes teardown more than once. I didn't really bother investigating why, it's not the end of the world to do so if it doesn't cause problems I don't think.

File: code/lib/preview-web/src/PreviewWeb.test.ts [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /// <reference types="@types/jest" />;
-
+import { jest, jest as mockJest, it, describe, beforeEach, afterEach, expect } from '@jest/globals';
```
Comment (shilman) [inline code/lib/preview-web/src/PreviewWeb.test.ts:2]: do we need this?

**Review by shilman — APPROVED** [review]
LGTM!

NOTE: CI is failing because this is based on an old version of `next` and we removed one of the templates due to test flake. Might revisit the implementation of that, but I think this run is OK for now.

File: code/lib/preview-web/src/render/StandaloneDocsRender.ts (hunk 1 above)
Comment (shilman) [reply code/lib/preview-web/src/render/StandaloneDocsRender.ts:102]: SGTM

File: code/lib/preview-web/src/PreviewWeb.test.ts (hunk 3 above)
Comment (tmeasday) [reply code/lib/preview-web/src/PreviewWeb.test.ts:2]: Strictly no, but I prefer it, rather than having global symbols available.

**Pull request merged**